scipy/sparse/lil.py

```diff
@@ -453,14 +453,23 @@ def tocsr(self, copy=False):
         lst = [len(x) for x in self.rows]
         idx_dtype = get_index_dtype(maxval=max(self.shape[1], sum(lst)))
         indptr = np.cumsum([0] + lst, dtype=idx_dtype)
-        indices = np.empty(indptr[-1], dtype=idx_dtype)
-        data = np.empty(indptr[-1], dtype=self.dtype)
-
-        start = 0
-        for stop, indices_i, data_i in zip(indptr[1:], self.rows, self.data):
-            indices[start:stop] = indices_i
-            data[start:stop] = data_i
-            start = stop
+        # choose faster construction approach depending on density 
+        if len(self.rows) == 0:
+            indices = np.empty(0, dtype=idx_dtype)
+            data = np.empty(0, dtype=self.dtype)            
+        elif indptr[-1] / len(self.rows) >= 25:
+            indices = np.empty(indptr[-1], dtype=idx_dtype)
+            data = np.empty(indptr[-1], dtype=self.dtype)
+            start = 0
+            for i, stop in enumerate(indptr[1:]):
+                if stop > start:
+                    indices[start:stop] = self.rows[i]
+                    data[start:stop] = self.data[i]
+                    start = stop
+        else:
+            indices = np.array([x for y in self.rows for x in y], dtype=idx_dtype)
+            data = np.array([x for y in self.data for x in y], dtype=self.dtype)
```

Comment:
We might be able to accelerate this case, too. I haven't benchmarked it, but what about:

```
nnz = indptr[-1]
indices = np.fromiter((x for y in self.rows for x in y), dtype=idx_dtype, count=nnz)
data = np.fromiter((x for y in self.data for x in y), dtype=self.dtype, count=nnz)
```

Reply:
I have run the same benchmarks (https://github.com/scipy/scipy/pull/10921#issuecomment-543108608) for the `fromiter` approach. It is faster than the old approach with `array`. But for more dense matrices, the approach with the `for` loop is still faster. Next are the results for the `fromiter` approach:

```
time = 0.0007870905000004313 for density = 0.5 and n = 100
time = 0.00019224930000518725 for density = 0.05 and n = 100
time = 0.00013232850000122198 for density = 0.005 and n = 100
time = 0.00017769400000133828 for density = 0.0005 and n = 100
time = 0.00024154620000444994 for density = 5e-05 and n = 100
time = 0.00027562819999502606 for density = 5e-06 and n = 100
time = 0.0005014370000026247 for density = 5e-07 and n = 100
time = 0.00010796990000017104 for density = 5e-08 and n = 100
time = 0.00021997100000135106 for density = 5e-09 and n = 100
time = 0.00017966780000051586 for density = 5e-10 and n = 100
time = 0.0004466785000090567 for density = 5e-11 and n = 100

time = 0.0604202551999947 for density = 0.5 and n = 1000
time = 0.0069927543000062545 for density = 0.05 and n = 1000
time = 0.0011015346000021963 for density = 0.005 and n = 1000
time = 0.0017074651000029916 for density = 0.0005 and n = 1000
time = 0.00038313959998959035 for density = 5e-05 and n = 1000
time = 0.00035206139999672815 for density = 5e-06 and n = 1000
time = 0.0006028711999988446 for density = 5e-07 and n = 1000
time = 0.0006456008999975893 for density = 5e-08 and n = 1000
time = 0.00026355769999781843 for density = 5e-09 and n = 1000
time = 0.000261667999996007 for density = 5e-10 and n = 1000
time = 0.00025958630000104675 for density = 5e-11 and n = 1000

time = 6.832772028900001 for density = 0.5 and n = 10000
time = 0.5872168207000072 for density = 0.05 and n = 10000
time = 0.08762907110000243 for density = 0.005 and n = 10000
time = 0.009471909199999119 for density = 0.0005 and n = 10000
time = 0.0038652721999937966 for density = 5e-05 and n = 10000
time = 0.004069516399999884 for density = 5e-06 and n = 10000
time = 0.0028417866000040705 for density = 5e-07 and n = 10000
time = 0.0028000212000051762 for density = 5e-08 and n = 10000
time = 0.0021166982999943685 for density = 5e-09 and n = 10000
time = 0.0017165158000011615 for density = 5e-10 and n = 10000
time = 0.0018980359000011048 for density = 5e-11 and n = 10000
```
and
```
time = 0.00011804600000004939 for density = 0.0 and n = 100
time = 0.0003789584999935869 for density = 0.1 and n = 100
time = 0.0008207838000089396 for density = 0.2 and n = 100
time = 0.0004918886000041311 for density = 0.3 and n = 100
time = 0.0014969394999980068 for density = 0.4 and n = 100
time = 0.0017561689999979536 for density = 0.5 and n = 100
time = 0.0013262999000062336 for density = 0.6 and n = 100
time = 0.0021951981000029263 for density = 0.7 and n = 100
time = 0.001845077500001935 for density = 0.8 and n = 100
time = 0.001547944000003554 for density = 0.9 and n = 100
time = 0.001961753000000499 for density = 1.0 and n = 100

time = 0.00026508299999932204 for density = 0.0 and n = 1000
time = 0.0016252742999995462 for density = 0.01 and n = 1000
time = 0.0029982689000007667 for density = 0.02 and n = 1000
time = 0.004226863799999592 for density = 0.03 and n = 1000
time = 0.0058504106999976106 for density = 0.04 and n = 1000
time = 0.007046701599995231 for density = 0.05 and n = 1000
time = 0.008153889300001538 for density = 0.06 and n = 1000
time = 0.009567921499990462 for density = 0.07 and n = 1000
time = 0.010691102199996295 for density = 0.08 and n = 1000
time = 0.012088687100003881 for density = 0.09 and n = 1000
time = 0.013735396699996728 for density = 0.1 and n = 1000

time = 0.0021405577999985327 for density = 0.0 and n = 10000
time = 0.0165631435000023 for density = 0.001 and n = 10000
time = 0.05708648510000103 for density = 0.002 and n = 10000
time = 0.04435130570000183 for density = 0.003 and n = 10000
time = 0.05566386140000077 for density = 0.004 and n = 10000
time = 0.06753021910000143 for density = 0.005 and n = 10000
time = 0.07805487489999904 for density = 0.006 and n = 10000
time = 0.09072074760000533 for density = 0.007 and n = 10000
time = 0.100841660399999 for density = 0.008 and n = 10000
time = 0.11972746789999747 for density = 0.009 and n = 10000
time = 0.12389069509999899 for density = 0.01 and n = 10000
```